File: frontend/src/api/annotations.js
/**
 * Annotations API module for PDF highlight/comment annotations.
 */

import { API_BASE, authFetch } from './client.js';

/**
 * Get all annotations for an attachment.
 * @param {string} attachmentId - The attachment UUID
 * @returns {Promise<Array>} List of annotation objects
 */
export async function getAnnotations(attachmentId) {
  const response = await authFetch(`${API_BASE}/attachments/${attachmentId}/annotations`);
  if (!response.ok) throw new Error('Failed to get annotations');
  return response.json();
}

/**
 * Add a new annotation to an attachment.
 * @param {string} attachmentId - The attachment UUID
 * @param {Object} data - Annotation data
 * @param {number} data.page - Page number
 * @param {string} data.type - Annotation type (highlight, comment, etc.)
 * @param {string} data.content - Annotation text content
 * @param {Array} data.rects - Array of {x, y, width, height} rectangles
 * @param {string} [data.color] - Highlight color (default #FFEB3B)
 * @returns {Promise<Object>} Created annotation
 */
export async function addAnnotation(attachmentId, { page, type, content, rects, color }) {
  const response = await authFetch(`${API_BASE}/attachments/${attachmentId}/annotations`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ page, type, content, rects, color }),
  });
  if (!response.ok) throw new Error('Failed to add annotation');
  return response.json();
}

/**
 * Update an existing annotation.
 * @param {string} attachmentId - The attachment UUID
 * @param {string} annotationId - The annotation UUID
 * @param {Object} data - Fields to update (content, color)
 * @returns {Promise<Object>} Updated annotation
 */
export async function updateAnnotation(attachmentId, annotationId, data) {
  const response = await authFetch(
    `${API_BASE}/attachments/${attachmentId}/annotations/${annotationId}`,
    {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    }
  );
  if (!response.ok) throw new Error('Failed to update annotation');
  return response.json();
}

/**
 * Delete an annotation.
 * @param {string} attachmentId - The attachment UUID
 * @param {string} annotationId - The annotation UUID
 * @returns {Promise<void>}
 */
export async function deleteAnnotation(attachmentId, annotationId) {
  const response = await authFetch(
    `${API_BASE}/attachments/${attachmentId}/annotations/${annotationId}`,
    { method: 'DELETE' }
  );
  if (!response.ok) throw new Error('Failed to delete annotation');
}

File: backend/routes/annotations.py
"""API routes for PDF annotations."""

import uuid
from typing import Optional, List, Dict, Any

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from ..database.connection import get_db
from ..database.models import User
from ..database import crud as db_crud
from ..auth.dependencies import get_current_user
from ..utils import parse_uuid

router = APIRouter(prefix="/api", tags=["annotations"])


# ---- Pydantic models ----

class RectModel(BaseModel):
    x: float
    y: float
    width: float
    height: float


class AddAnnotationRequest(BaseModel):
    page: int = Field(..., ge=0)
    type: str = Field(default="highlight")
    content: str = Field(default="")
    rects: List[RectModel] = Field(default_factory=list)
    color: str = Field(default="#FFEB3B")


class UpdateAnnotationRequest(BaseModel):
    content: Optional[str] = None
    color: Optional[str] = None


class AnnotationResponse(BaseModel):
    id: str
    page: int
    type: str
    content: str
    rects: List[Dict[str, Any]]
    color: str
    created_at: str

    class Config:
        from_attributes = True


# ---- Endpoints ----

@router.get(
    "/attachments/{attachment_id}/annotations",
    response_model=List[AnnotationResponse],
)
async def list_annotations(
    attachment_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get all annotations for an attachment."""
    annotations = await db_crud.annotations.get_annotations(
        db, parse_uuid(attachment_id, "attachment_id")
    )
    return annotations


@router.post(
    "/attachments/{attachment_id}/annotations",
    response_model=AnnotationResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_annotation(
    attachment_id: str,
    request: AddAnnotationRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Add a new annotation to an attachment."""
    try:
        annot = await db_crud.annotations.add_annotation(
            db,
            attachment_id=parse_uuid(attachment_id, "attachment_id"),
            page=request.page,
            type=request.type,
            content=request.content,
            rects=[r.model_dump() for r in request.rects],
            color=request.color,
        )
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    await db.commit()
    return annot


@router.patch(
    "/attachments/{attachment_id}/annotations/{annotation_id}",
    response_model=AnnotationResponse,
)
async def update_annotation(
    attachment_id: str,
    annotation_id: str,
    request: UpdateAnnotationRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Update an existing annotation."""
    updated = await db_crud.annotations.update_annotation(
        db,
        attachment_id=parse_uuid(attachment_id, "attachment_id"),
        annotation_id=annotation_id,
        content=request.content,
        color=request.color,
    )
    if updated is None:
        raise HTTPException(status_code=404, detail="Annotation not found")
    await db.commit()
    return updated


@router.delete(
    "/attachments/{attachment_id}/annotations/{annotation_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_annotation(
    attachment_id: str,
    annotation_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Delete an annotation."""
    deleted = await db_crud.annotations.delete_annotation(
        db,
        attachment_id=parse_uuid(attachment_id, "attachment_id"),
        annotation_id=annotation_id,
    )
    if not deleted:
        raise HTTPException(status_code=404, detail="Annotation not found")
    await db.commit()
